modules/picard_qc/1.0/picard_qc.smk

```diff
@@ -0,0 +1,323 @@
+)
+
+# Define rules to be run locally when using a compute cluster
+localrules:
+    _picard_qc_input_bam,
+    _picard_qc_rrna_int,
+    _picard_qc_merge_metrics,
+    _picard_qc_merged_output,
+    _picard_qc_flagstats_output,
+    _picard_qc_merged_dispatch,
+    _picard_qc_all
+
+
+##### RULES #####
+
+
+# Symlinks the input files into the module results directory (under '00-inputs/')
+rule _picard_qc_input_bam:
+    input:
+        bam = CFG["inputs"]["sample_bam"]
+    output:
+        bam = CFG["dirs"]["inputs"] + "{seq_type}--{genome_build}/{sample_id}.bam"
+    run:
+        op.relative_symlink(input.bam, output.bam)
+
+
+rule _picard_qc_alignment_summary:
+    input:
+        bam = rules._picard_qc_input_bam.output.bam,
+        fasta = reference_files("genomes/{genome_build}/genome_fasta/genome.fa")
+    output:
+        metrics = CFG["dirs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/alignment_summary_metrics",
+    log:
+        stdout = CFG["logs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/alignment_summary.stdout.log",
+        stderr = CFG["logs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/alignment_summary.stderr.log"
+    params:
+        opts = CFG["options"]["alignment_summary"]
+    conda:
+        CFG["conda_envs"]["picard"]
+    threads:
+        CFG["threads"]["alignment_summary"]
+    resources:
+        mem_mb = CFG["mem_mb"]["alignment_summary"]
+    shell:
+        op.as_one_line("""
+        picard -Xmx{resources.mem_mb}m CollectAlignmentSummaryMetrics
+        {params.opts} 
+        I={input.bam} O={output.metrics} R={input.fasta} 
+        > {log.stdout} 2> {log.stderr}
+        """)
+
+
+rule _picard_qc_insert_size:
+    input:
+        bam = rules._picard_qc_input_bam.output.bam,
+    output:
+        metrics = CFG["dirs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/insert_size_metrics",
+        histogram = CFG["dirs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/insert_size_histogram.pdf"
+    log:
+        stderr = CFG["logs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/insert_size.stderr.log",
+        stdout = CFG["logs"]["metrics"] + "{seq_type}--{genome_build}/{sample_id}/insert_size.stdout.log"
+    params:
+        opts = CFG["options"]["insert_size"]
+    conda:
+        CFG["conda_envs"]["picard"]
+    threads:
+        CFG["threads"]["insert_size"]
+    resources:
+        mem_mb = CFG["mem_mb"]["insert_size"]
+    shell:
+        op.as_one_line("""
+        picard -Xmx{resources.mem_mb}m CollectInsertSizeMetrics
+        {params.opts} 
+        I={input.bam} O={output.metrics} H={output.histogram}
+        > {log.stdout} 2> {log.stderr}
+        """)
+
+# capture metrics
+rule _picard_qc_hs_metrics:
+    input:
+        bam = rules._picard_qc_input_bam.output.bam,
+        fasta = reference_files("genomes/{genome_build}/genome_fasta/genome.fa"),
+        intervals = CFG["inputs"]["intervals"]
```

Comment:
Because it's so easy to generate an interval_list file from a BED file and BAM file, could we not include a rule for this here since we already are using picard? See example command below:

```
picard BedToIntervalList I=capture_regions.bed O=capture_regions.interval_list SD=BLGSP-71-06-00080-01A-11D.bam
```

Reply:
Ya we can, so users need to provide a bed file instead of an intervals file?

Reply:
I forgot I wrote that rule in the `utils` module

Reply:
Would this solution mean that a separate intervals file is generated for each BAM?

Reply:
no I don't think so, it's only generated once for each bed file, since the interval files are created in a `reference` directory which is outside of the `result` directory